lib/connection_controller.go

```diff
@@ -231,6 +252,84 @@ func (cc *ConnectionController) refreshConnectIps() {
 	}
 }
 
+// ###########################
+// ## Validator Connections
+// ###########################
+
+// refreshValidatorIndex re-indexes validators based on the activeValidatorsMap. It is called periodically by the
+// validator connector.
+func (cc *ConnectionController) refreshValidatorIndex(activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, *ValidatorEntry]) {
+	// De-index inactive validators. We skip any checks regarding RemoteNodes connection status, nor do we verify whether
+	// de-indexing the validator would result in an excess number of outbound/inbound connections. Any excess connections
+	// will be cleaned up by the peer connector.
+	validatorRemoteNodeMap := cc.rnManager.GetValidatorIndex().Copy()
+	for pk, rn := range validatorRemoteNodeMap {
+		// If the validator is no longer active, de-index it.
+		if _, ok := activeValidatorsMap.Get(pk); !ok {
+			cc.rnManager.SetNonValidator(rn)
+			cc.rnManager.UnsetValidator(rn)
+		}
+	}
+
+	// Look for validators in our existing outbound / inbound connections.
+	allNonValidators := cc.rnManager.GetAllNonValidators()
+	for _, rn := range allNonValidators {
+		// It is possible for a RemoteNode to be in the non-validator indices, and still have a public key. This can happen
+		// if the RemoteNode advertised support for the SFValidator service flag during handshake, and provided us
+		// with a public key, and a corresponding proof of possession signature.
+		pk := rn.GetValidatorPublicKey()
+		if pk == nil {
+			continue
+		}
+		// It is possible that through unlikely concurrence, and malevolence, two non-validators happen to have the same
+		// public key, which goes undetected during handshake. To prevent this from affecting the indexing of the validator
+		// set, we check that the non-validator's public key is not already present in the validator index.
+		if _, ok := cc.rnManager.GetValidatorIndex().Get(pk.Serialize()); ok {
+			cc.rnManager.Disconnect(rn)
+			continue
+		}
+
+		// If the RemoteNode turns out to be in the validator set, index it.
+		if _, ok := activeValidatorsMap.Get(pk.Serialize()); ok {
+			cc.rnManager.SetValidator(rn)
+			cc.rnManager.UnsetNonValidator(rn)
+		}
+	}
+}
+
+// connectValidators attempts to connect to all active validators that are not already connected. It is called
+// periodically by the validator connector.
+func (cc *ConnectionController) connectValidators(activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, *ValidatorEntry]) {
+	// Look through the active validators and connect to any that we're not already connected to.
+	if cc.blsKeystore == nil {
+		return
+	}
+
+	validators := activeValidatorsMap.Copy()
+	for pk, validator := range validators {
+		_, exists := cc.rnManager.GetValidatorIndex().Get(pk)
+		// If we're already connected to the validator, continue.
+		if exists {
+			continue
+		}
+		if cc.blsKeystore.GetSigner().GetPublicKey().Serialize() == pk {
```

Comment:
We should use bytes.Equal here.

Reply:
I think this one is okay to leave as `==`, both types are strings actually.